crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,359 @@ impl<'a> FileNoqaDirectives<'a> {
+        while !self.cursor.is_eof() {
+            self.lex_code()?;
+        }
+        Ok(())
+    }
+
+    fn lex_code(&mut self) -> Result<(), LexicalError> {
+        self.cursor.start_token();
+        match self.cursor.first() {
+            c if c.is_ascii_whitespace() => {
+                self.cursor.eat_while(|chr| chr.is_ascii_whitespace());
+                // Ex) # noqa:    ,F401
+                //            ^^^^^
+                if self.cursor.first() == ',' {
+                    self.warnings.push(LexicalWarning::MissingItem(
+                        self.token_range().add(self.offset),
+                    ));
+                    self.cursor.eat_char(',');
+                }
+            }
+            // Ex) # noqa: F401,,F841
+            //                  ^
+            ',' => {
+                self.warnings.push(LexicalWarning::MissingItem(
+                    self.token_range().add(self.offset),
+                ));
+                self.cursor.eat_char(',');
+            }
+
+            // Ex) # noqa: F401
+            //             ^
+            c if c.is_ascii_uppercase() => {
+                self.cursor.eat_while(|chr| chr.is_ascii_uppercase());
+                if !self.cursor.first().is_ascii_digit() {
+                    // Fail hard if we're already attempting
+                    // to lex squashed codes, e.g. `F401Fabc`
+                    if self.missing_delimiter {
+                        return Err(LexicalError::InvalidCodeSuffix);
+                    }
+                    // Otherwise we've reached the first invalid code,
+                    // so it could be a comment and we just stop parsing.
+                    // Ex) `#noqa: F401 A comment`
+                    //        we're here^
+                    self.cursor.skip_bytes(self.cursor.as_str().len());
+                    return Ok(());
+                }
+                self.cursor.eat_while(|chr| chr.is_ascii_digit());
+                self.push_code();
+
+                if self.cursor.is_eof() {
+                    return Ok(());
+                }
+
+                self.missing_delimiter = match self.cursor.first() {
+                    ',' => {
+                        self.cursor.eat_char(',');
+                        false
+                    }
+                    c if c.is_ascii_whitespace() => false,
+                    c if c.is_ascii_uppercase() => {
+                        self.warnings
+                            .push(LexicalWarning::MissingDelimiter(TextRange::empty(
+                                self.offset + self.current(),
+                            )));
+                        true
+                    }
+                    '#' => false,
+                    _ => return Err(LexicalError::InvalidCodeSuffix),
+                };
+            }
+            _ => {
+                self.cursor.skip_bytes(self.cursor.as_str().len());
```

Comment:
Should this push an error, considering that it isn't a valid code?

Reply:
Let's add a few comments with examples to those match arms. I would find them useful :)